### ISSUE TITLE
Clarify User-Code Interaction

### ISSUE THREAD

**Issue by yaronf** [ietf-wg-gnap/gnap-core-protocol#9]
Yaron: it is unclear to me what the goal of the user-code interaction is, in other words what is the threat model. If it's just "second factor, something you have", then an OTP MFA device would be simpler and offer the same security.

Justin: This is an equivalent function of the OAuth Device Flow: https://tools.ietf.org/html/rfc8628

It’s not about a second factor for authenticating the user, really, it’s about facilitating authorization on a secondary device through the use of a user-typed code to tie the back end and front end together.

Tom Jones: Ok - several others have been asking as well.  I wrote up this page https://tcwiki.azurewebsites.net/index.php?title=Apps_on_User_Devices

basically there are assertions from the programmer of the app, and from the app itself that are passed in the id token (whatever) from the user device.
Note that this means that not all assertions are signed by the same key in this case!
Nor are all the attributes about the subject. Some are just in support of the level of assurance.
The big difference with FIDO U2F and II is that the keys are bound to the user in the device, not in the RP.

**Comment by TomCJones** [comment]
The issue is about the provenance of the user assertions.  Or - how can the recipient know that code faithfully represents the user for both intent and presence. In other words, if the provenance of the 2nd factor is in doubt, why should it be trusted. I would point out the web authN addresses this issue and well as payment request in W3C.

**Comment by jricher** [comment]
The user code is an OTP issued by the AS to identify the client's request, not to identify the user. It's assumed that the user will authenticate at the AS on top of presenting the user code. That user authentication could use MFA of some type, entirely separate from the user code, which is where webauthn and FIDO would come into play.

The precedent for this feature is the OAuth device flow where the user uses a separate device to interact with the AS and authorize the client. The device flow, like OAuth itself, doesn't say anything about how the user authenticates.

**Comment by TomCJones** [comment]
OK, that is your model. In my model all of the IAL2 and AAL2 claims get bundled up into a single packet and sent to the AS and GC. I know that GNAP doesn't claim to be OIDC, but it should at least co-exist with it.

**Comment by fimbault** [comment]
"GNAP doesn't claim to be OIDC, but it should at least co-exist with it." yes indeed, but that doesn't mean we need to bundle everything at the core of GNAP. We can also integrate with whatever identity layer is chosen (OIDC, SSI, etc) as extension points.
As a first step it would be useful to make a complete example of GNAP + OIDC working together.

As for the clarification requested in this thread, the separation of concerns makes it clearer I think. The user code isn't MFA.

**Comment by jricher** [comment]
Need to explain in the user code section (or sections) that the code isn't MFA for the user, it's meant to associate the user's session with the pending grant request. See also proposed changes in #123
